### PR TITLE
feat: Use docker or podman to build and run

### PR DESCRIPTION
**Pull request by rmarting**
This PR improves the Makefile to check which container tool is installed in the enviroment (only docker or podman).

The user will not have to set up anything before to build or to run the image locally, only a container tool is needed to be installed.

HTH

### CODE REVIEW

**Comment by rmarting** [comment]
@ashokponkumar Here the PR as you request me related with https://github.com/konveyor/move2kube-ui/pull/98

Ready for your review.

**Comment by rmarting** [comment]
@HarikrishnanBalagopal I don't know why DCO is not passing. I followed the instructions to signoff my latest commit, but it is something not working.

From my local repo:

```shell
move2kube-api on  use-container-tool via 🐹 v1.16.8 via ⬢ v10.17.0 took 8s 
❯ git commit --amend --no-edit --signoff
[use-container-tool bfa4189] :twisted_rightwards_arrows: Merge branch 'use-container-tool' of https://github.com/rmarting/move2kube-api into use-container-tool
 Date: Thu Sep 30 12:06:19 2021 +0200
move2kube-api on  use-container-tool via 🐹 v1.16.8 via ⬢ v10.17.0 
❯ git push --force-with-lease origin use-container-tool
Enumerating objects: 1, done.
Counting objects: 100% (1/1), done.
Writing objects: 100% (1/1), 311 bytes | 311.00 KiB/s, done.
Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/rmarting/move2kube-api.git
 + c4d900a...bfa4189 use-container-tool -> use-container-tool (forced update)
```

What am I doing wrong?

**Comment by ashokponkumar** [comment]
> @HarikrishnanBalagopal I don't know why DCO is not passing. I followed the instructions to signoff my latest commit, but it is something not working.
> 
> From my local repo:
> 
> ```shell
> move2kube-api on  use-container-tool via 🐹 v1.16.8 via ⬢ v10.17.0 took 8s 
> ❯ git commit --amend --no-edit --signoff
> [use-container-tool bfa4189] :twisted_rightwards_arrows: Merge branch 'use-container-tool' of https://github.com/rmarting/move2kube-api into use-container-tool
>  Date: Thu Sep 30 12:06:19 2021 +0200
> move2kube-api on  use-container-tool via 🐹 v1.16.8 via ⬢ v10.17.0 
> ❯ git push --force-with-lease origin use-container-tool
> Enumerating objects: 1, done.
> Counting objects: 100% (1/1), done.
> Writing objects: 100% (1/1), 311 bytes | 311.00 KiB/s, done.
> Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
> To https://github.com/rmarting/move2kube-api.git
>  + c4d900a...bfa4189 use-container-tool -> use-container-tool (forced update)
> ```
> 
> What am I doing wrong?

Possibly one of the commits ( I guess the merge commit) does not have the sign off. Possibly you can squash and sign that commit.

**Comment by rmarting** [comment]
> Possibly one of the commits ( I guess the merge commit) does not have the sign off. Possibly you can squash and sign that commit.

Squash some commits into single one, but now this PR includes my changes and others coming from the main branch. TBH I don't know what it is happeing, but at least the commits include my changes.

Ready for your review.

**Comment by ashokponkumar** [comment]
> > Possibly one of the commits ( I guess the merge commit) does not have the sign off. Possibly you can squash and sign that commit.
> 
> Squash some commits into single one, but now this PR includes my changes and others coming from the main branch. TBH I don't know what it is happeing, but at least the commits include my changes.
> 
> Ready for your review.
> 

Can you please try a rebase with main? It shows a warning that the branch is out of date.

**Comment by rmarting** [comment]
Done!

File: Makefile [hunk 1]
```diff
@@ -190,21 +200,43 @@ info: ## Get version info
...
+ifndef CONTAINER_TOOL
+$(error No container tool (docker, podman) found in your environment. Please, install one)
+endif
+
+	@echo "Running image with $(CONTAINER_TOOL)"
+
+ifdef DOCKER_CMD
+	${CONTAINER_TOOL} run --rm -p 8080:8080 -v /var/run/docker.sock:/var/run/docker.sock -v ${PWD}/:/workspace ${REGISTRYNS}/${BINNAME}:${VERSION}
+else
+	${CONTAINER_TOOL} run --rm -it -p 8080:8080 --network=bridge ${REGISTRYNS}/${BINNAME}:${VERSION}
```
Comment (ashokponkumar) [inline Makefile:241]: @rmarting Do we need to run with -it flag for it to work?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Just a nit question.

File: Makefile [hunk 2]
```diff
@@ -190,21 +200,43 @@ info: ## Get version info
...
-	docker push ${REGISTRYNS}/${BINNAME}:${VERSION}
+	${CONTAINER_TOOL} push ${REGISTRYNS}/${BINNAME}-builder:${VERSION}
+	${CONTAINER_TOOL} push ${REGISTRYNS}/${BINNAME}:${VERSION}
 
 .PHONY: crun
-crun: ## Run docker image
-	docker run --rm -p 8080:8080 -v /var/run/docker.sock:/var/run/docker.sock -v ${PWD}/:/workspace ${REGISTRYNS}/${BINNAME}:${VERSION}
+crun: ## Run container image
+ifndef CONTAINER_TOOL
+$(error No container tool (docker, podman) found in your environment. Please, install one)
```
Comment (ashokponkumar) [inline Makefile:233]: This line needs a tab space at start

File: Makefile [hunk 3]
```diff
@@ -190,21 +200,43 @@ info: ## Get version info
...
+	${CONTAINER_TOOL} build -t ${REGISTRYNS}/${BINNAME}-builder:${VERSION} --cache-from ${REGISTRYNS}/${BINNAME}-builder:latest --target build_base                          --build-arg VERSION=${VERSION} --build-arg GO_VERSION=${GO_VERSION} .
+	${CONTAINER_TOOL} build -t ${REGISTRYNS}/${BINNAME}:${VERSION}         --cache-from ${REGISTRYNS}/${BINNAME}-builder:latest --cache-from ${REGISTRYNS}/${BINNAME}:latest --build-arg VERSION=${VERSION} --build-arg GO_VERSION=${GO_VERSION} .
+	${CONTAINER_TOOL} tag ${REGISTRYNS}/${BINNAME}-builder:${VERSION} ${REGISTRYNS}/${BINNAME}-builder:latest
+	${CONTAINER_TOOL} tag ${REGISTRYNS}/${BINNAME}:${VERSION} ${REGISTRYNS}/${BINNAME}:latest
 
 .PHONY: cpush
-cpush: ## Push docker image
+cpush: ## Push container image
+ifndef CONTAINER_TOOL
+$(error No container tool (docker, podman) found in your environment. Please, install one)
```
Comment (ashokponkumar) [inline Makefile:221]: This line needs a tab space at start

File: Makefile [hunk 4]
```diff
@@ -190,21 +200,43 @@ info: ## Get version info
...
 
 .PHONY: cbuild
-cbuild: ## Build docker image
-	docker build -t ${REGISTRYNS}/${BINNAME}-builder:${VERSION} --cache-from ${REGISTRYNS}/${BINNAME}-builder:latest --target build_base                          --build-arg VERSION=${VERSION} --build-arg GO_VERSION=${GO_VERSION} .
-	docker build -t ${REGISTRYNS}/${BINNAME}:${VERSION}         --cache-from ${REGISTRYNS}/${BINNAME}-builder:latest --cache-from ${REGISTRYNS}/${BINNAME}:latest --build-arg VERSION=${VERSION} --build-arg GO_VERSION=${GO_VERSION} .
-	docker tag ${REGISTRYNS}/${BINNAME}-builder:${VERSION} ${REGISTRYNS}/${BINNAME}-builder:latest
-	docker tag ${REGISTRYNS}/${BINNAME}:${VERSION} ${REGISTRYNS}/${BINNAME}:latest
+cbuild: ## Build container image
+ifndef CONTAINER_TOOL
+$(error No container tool (docker, podman) found in your environment. Please, install one)
```
Comment (ashokponkumar) [inline Makefile:208]: This line needs a tab space at start

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Even within `if` block, the lines need to be tab spaced at start, otherwise it seems to get executed irrespective of whether the section is invoked or not.

Just a nit, just to make each section standout, we have refrained in general from using empty lines within each section in the makefile. If you think its okay, we can remove the empty line within the section.

File: Makefile (hunk 1 above)
Comment (rmarting) [reply Makefile:241]: No, we don't. I use it normally but TBH for this use case it could be not needed.

File: Makefile (hunk 4 above)
Comment (rmarting) [reply Makefile]: Done!

File: Makefile (hunk 3 above)
Comment (rmarting) [reply Makefile]: Done!

File: Makefile (hunk 2 above)
Comment (rmarting) [reply Makefile]: Done!

File: Makefile (hunk 1 above)
Comment (ashokponkumar) [reply Makefile:241]: Would be good to remove it, since the api container will be running in non interactive mode.

Comment (rmarting) [reply Makefile]: It makes sense. Done!

**Review by ashokponkumar — APPROVED** [review]

**Pull request merged**